machine/v1beta1/types_azureprovider.go

```diff
@@ -397,6 +419,28 @@ type OSDiskManagedDiskParameters struct {
 	// DiskEncryptionSet is the disk encryption set properties
 	// +optional
 	DiskEncryptionSet *DiskEncryptionSetParameters `json:"diskEncryptionSet,omitempty"`
+	// SecurityProfile specifies the security profile for the managed disk.
+	// +optional
+	SecurityProfile *VMDiskSecurityProfile `json:"securityProfile,omitempty"`
+}
+
+// VMDiskSecurityProfile specifies the security profile settings for the managed disk.
+// It can be set only for Confidential VMs.
+type VMDiskSecurityProfile struct {
+	// DiskEncryptionSet specifies the customer managed disk encryption set resource id for the
+	// managed disk that is used for Customer Managed Key encrypted ConfidentialVM OS Disk and
+	// VMGuest blob.
+	// +optional
+	DiskEncryptionSet *DiskEncryptionSetParameters `json:"diskEncryptionSet,omitempty"`
+	// SecurityEncryptionType specifies the encryption type of the managed disk.
+	// It is set to DiskWithVMGuestState to encrypt the managed disk along with the VMGuestState
+	// blob, and to VMGuestStateOnly to encrypt the VMGuestState blob only.
+	// When set to VMGuestStateOnly, the vTPM should be enabled.
+	// When set to DiskWithVMGuestState, both SecureBoot and vTPM should be enabled.
```

Comment:
What happens if these don't get enabled?

Reply:
[Security encryption type](https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?tabs=HTTP#securityencryptiontypes) can only be set for `SecurityType: ConfidentialVM`. When `vTPM` is disabled (in the case of `VMGuestStateOnly`) or both `vTPM` and `SecureBoot` are disabled (in the case of `DiskWithVMGuestState`, the Azure REST API will return a bad request error and the VM won't get created.

Unfortunately, this is not directly mentioned in the Azure REST API documentation, but the [Azure Terraform provider documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/windows_virtual_machine#security_encryption_type) and [implementation](https://github.com/hashicorp/terraform-provider-azurerm/blob/ed8e0b6508d69b39fa335d8ecd40061bc71adb0c/internal/services/compute/linux_virtual_machine_resource.go#L541-L561) includes this part.

Reply:
Hmm, so the disk security is not tied to the VM in the API, that's a shame. In upstream, how does this look? Vaguely similar to this? Wondering if there was a way to tie these params into a discriminated union to you could only set them once you'd set the relevant other options

Reply:
That's correct. According to the Azure REST API, although the OS Disk security and the VM security are tightly coupled, they reside in separate sections. 

The upstream [implementation](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L635-L665) follows the [respective](https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?tabs=HTTP#create-a-vm-with-securitytype-confidentialvm-with-customer-managed-keys) Azure REST API as closely as possible.

IMHO if we diverge significantly from the Azure REST API, we could end up with breaking changes in the future. In addition, although the Azure REST API is not intuitive at first sight, the documentation and tutorials for ConfidentialVMs are quite extensive and detailed. So I would definitely stick to the OpenShift API conventions as much as possible, but maybe compromise in terms of the structures we are introducing.  WDYT?

Reply:
Ok, in this case I think you're right, lets stick to having the split, but we should make sure the godoc explains what is going on and where to find the related issues

Reply:
Absolutely, I'll try to improve and refine the godoc of those fields and update the PR.

PS. We will also include all relevant validations of those fields and their combinations in the [Machine API side](https://github.com/openshift/machine-api-provider-azure/pull/54/files#diff-d25ff6792ca687a188e54899ac47fe10e9c11fcb14219ccb16f762cd6bd38140R454-R545) (and the respective Machine API operator webhook, this PR has not yet been opened, as soon as we have review/feedback on the MAPI side I'll go for the webhook PR as well).

Reply:
@JoelSpeed I have updated the godoc to reflect as many details as possible and a link to the official Azure docs on Confidential VMs. Please let me know you think this is enough or we should include or edit anything else. Thank you.